src/streaming-client.ts

```
@@ -0,0 +1,157 @@
+        this.client = new sockets.client();
+    }
+
+    /**
+     * Sets up the client and begins the streaming connection. Returns a duplex
+     * from which the user can read responses from the api and to which the user
+     * should write their audio data
+     *
+     * @returns BufferedDuplex. Data written to this buffer will be sent to the api
+     * Data received from the api can be read from this duplex
+     */
+    start(): Duplex {
+        this.setUpClient();
+        this.client.connect(this.baseUrl);
+        return new BufferedDuplex(this.requests, this.responses, { readableObjectMode: true });
+    }
+
+    /**
+     * Ends the streaming connection and closes off the buffer returned from start()
+     */
+    end(): void {
+        this.client.abort();
+        this.requests.end('EOS');
+        this.responses.push(null);
+    }
+
+    private setUpClient(): void {
+        this.setUpHttpResponseHandler();
+        this.setUpConnectionFailureHandler();
+        this.setUpConnectedHandlers();
+    }
+
+    private setUpHttpResponseHandler(): void {
+        this.client.on('httpResponse', (response: any) => {
+            this.emit('httpResponse', response.statusCode);
+            this.end();
+        });
+    }
+
+    private setUpConnectionFailureHandler(): void {
+        this.client.on('connectFailed', (error: Error) => {
+            this.emit('connectFailed', error);
+            this.end();
+        });
+    }
+
+    private setUpConnectedHandlers(): void {
+        this.client.on('connect', (connection: any) => {
+            connection.on('error', (error: Error) => {
+                this.emit('error', error);
+                this.end();
+            });
+            connection.on('close', (code: number, reason: string) => {
+                this.emit('close', code, reason);
+                this.end();
+            });
+            connection.on('message', (message: any) => {
+                if (message.type === 'utf8') {
```

what should happen if the message type is not utf8?

---

Why do we make this call?

---

I make this call to avoid items which are gauranteed to throw if we try to cast them as Hypotheses. I could log something on non utf8 messages but I don't think we should throw an error

---

I was mostly curious what non utf8 messages we were being sent.